packages/eslint-plugin/src/rules/require-object-type-annotations.ts

```diff
@@ -0,0 +1,54 @@
+import type { RuleListener } from '@typescript-eslint/utils/src/ts-eslint';
+import type * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'require-object-type-annotations',
+  meta: {
+    docs: {
+      description:
+        'Require type annotations for objects where there is no contextual type.',
+      requiresTypeChecking: true,
+      recommended: false,
+    },
+    messages: {
+      forbidden: 'Object is missing type annotation.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  defaultOptions: [],
+  create: (context): RuleListener => {
+    const parserServices = util.getParserServices(context);
+    const checker = parserServices.program.getTypeChecker();
+
+    return {
+      ObjectExpression: (esNode): void => {
```

Comment:
Here's one thing I need help with:

I have this test case which currently produces 2 errors:

```ts
const xs = { ys: [{ prop: 1 }] };
```

This makes sense because we have two `ObjectExpression`s that have no contextual type.

However, I think it would be better if we only reported one error for the outermost `ObjectExpression`. I thought this might be possible if I could tell ESLint to stop traversing, but it seems like that's not possible: https://github.com/typescript-eslint/typescript-eslint/issues/2734. With that in mind, is there any other way we could do this? Perhaps there's another rule that is doing something similar which I can take a look at.

Reply:
depends what the goal is for the inner object.
what is the plan for the user to "solve" this inner object?
with an `as` assertion? (I hope not!)
as that's the only solution - then you probably want to just ignore it entirely.

see my top level comment

Reply:
> with an as assertion? (I hope not!)

Certainly not 😛 

> what is the plan for the user to "solve" this inner object?

Once they annotate the outer object, the inner object will have a contextual type, meaning that language server features will for properties on both the inside and outside objects i.e. "rename" / "go to definition" / "find references"). See the video below where where I've tried to demonstrate this.

As I described in https://github.com/typescript-eslint/typescript-eslint/issues/3408, the goal of this rule is to make sure TypeScript understands which named type an object relates to, so that tools such as "go to definition" / "find references" / "rename" on object properties work properly and to avoid bugs

https://user-images.githubusercontent.com/921609/190969137-70454e8f-9634-4ebf-a9cd-aed9b6abc404.mov

Reply:
Here are some test cases for me to consider. Saving here so I don't lose it!

```ts
{
    declare const fn: <T>(t: T) => T;
    // 2 errors
    const x: { prop: { prop: number } } = { prop: fn({ prop: 1 }) };
}

{
    declare const fn: (t: { prop: number }) => unknown;
    // 1 error
    const x = { prop: fn({ prop: 1 }) };
}
{
    // 1 error
    const x = { prop: { prop: 1 } };
}
{
    // 1 error
    const xs = { ys: [{ prop: 1 }] };
}
{
    // 2 errors
    const x = {
        prop: () => {
            const y = { prop: 1 };
            return y;
        },
    };
}
{
    // 1 error
    const x = {
        prop: () => ({ prop: 1 }),
    };
}

```